### DISCUSSION TITLE
Pytest fixtures ordering in allure report seems random

### DISCUSSION THREAD

**Discussion by toitoinou** [allure-framework/allure2#2643] [Questions & Support] [answered]
Hello all,
I'm evaluating allure-pytest for our testing purpose, for which we need good reporting and procedures.

I'm trying to use the fixtures feature of pytest, here an example of what I intend to do:

```
import allure
import pytest


@pytest.fixture
def instanciate_mocks():
    """
    Setup actions, e.g., establishing a connection
    """
    mock = {"instanciate": True}
    yield mock
    mock["instanciate"] = False


#  Fixture library could be in a separated module
@pytest.fixture
def connect_mocks(instanciate_mocks):
    """
    Setup actions, e.g., establishing a connection
    """
    instanciate_mocks["connect"] = True
    yield instanciate_mocks
    instanciate_mocks["connect"] = False


@pytest.fixture
def put_mocks_in_phase(connect_mocks):
    """
    Setup actions, e.g., establishing a connection
    """
    connect_mocks["phase"] = True
    yield connect_mocks
    connect_mocks["phase"] = False


# Test script
class TestClass:
    @allure.step("Step 1")
    def step1(self):
        x = "this"
        assert "h" in x, "'h' was not found in 'this'"

    @allure.step("Step 2")
    def step2(self):
        x = "hello"
        assert "h" in x, "'h' was not found in 'hello'"

    def test_1(self, put_mocks_in_phase):
        """
        Test with steps and fixtures
        """
        self.mocks = put_mocks_in_phase
        self.step1()
        self.step2()
```
The test run properly, but in the report the set-up order seems to be random :
First run :
![image](https://github.com/user-attachments/assets/202eb300-87c9-4557-a6f2-80427f684ce7)

Second run :
![image](https://github.com/user-attachments/assets/5a42e11a-fde6-44b0-88d9-6c477b37d91c)

Is this a known issue or do I made a mistake during my environment setup ?

We use python 3.10.8 with allure 2.30.0, but the same issue occured on allure 2.29.

**Comment by delatrie** [comment, marked as answer]
Hi, @toitoinou ! Allure Report orders fixtures by their start time, which has a precision of 1 ms. If your fixtures execute fast enough, their start time is the same resulting in a random order. It can't be fixed until some notion of order is introduced into Allure's handling of fixtures.

As a workaround, you may introduce a small delay:

```python
import time

@pytest.fixture
def instanciate_mocks():
    """
    Setup actions, e.g., establishing a connection
    """
    mock = {"instanciate": True}
    time.sleep(0.001)
    yield mock
    time.sleep(0.001)
    mock["instanciate"] = False
```

It's not elegant, but solves the problem.

**Reply by toitoinou** [reply]
Thank you for your fast reply and your insight of how allure work.
